bokeh/embed.py

```diff
@@ -131,10 +154,27 @@ def components(models, wrap_script=True, wrap_plot_info=True):
             values.append(models[k])
         models = values
 
-    # 2) Do our rendering
+    # 2) Append models to one document. Either pre-existing or new.
+    existing_docs = set(m if isinstance(m, Document) else m.document for m in models)
+    existing_docs.discard(None)
 
-    with _ModelInDocument(models):
-        (docs_json, render_items) = _standalone_docs_json_and_render_items(models)
+    if len(existing_docs) == 0:
+        # no existing docs, make a new one
+        doc = Document()
+    elif len(existing_docs) == 1:
+        # all existing docs are the same, use that one
+        doc = existing_docs.pop()
+    else:
+        # conflicting/multiple docs, raise an error
+        msg = ('Multiple items in models conatain documents or are '
+               'themselves documents. (Models must be owned by only a '
+               'single document). This may indicate a usage error.')
+        raise RuntimeError(msg)
+
+    models_to_dedoc = list(_add_doc_to_models(doc, models))
```

Comment:
I think having `_add_doc_to_models` be a generator in this specific instance is a bit dangerous. At least, I would assume that a single call to `_add_doc_to_models` would add the doc to all the models. But it *won't* do that that, unless it happens to be used in a loop, or the caller remembers to wrap with `list`. I think it would be better for it to just return the list of models, not the generator.